pkg/generate/app/cmd/newapp.go

```diff
@@ -418,6 +418,19 @@ func (c *AppConfig) buildPipelines(components app.ComponentReferences, environme
 						return nil, err
 					}
 				}
+				// Append any exposed ports from Dockerfile to input image
```

Comment:
> Having this change here means that even after Resolve is finished, the ComponentInput is incomplete. I think the ComponentInput is getting mutated here via series of navigated references, but I'm not completely certain. If so, it feels a little weird to have something called buildPipelines change an input object.

Not quite. So the input image ref is created here in this method (using ref.Input().Match as the input) https://github.com/csrwng/origin/blob/newapp_dockerfile_ports/pkg/generate/app/cmd/newapp.go#L403, and we're simply mutating that created imageref.

Having said that, I agree that it is hard to follow what objects are available at what time. A restructuring and maybe a refactor to a more functional style of code here with immutable inputs/outputs would be great. But that won't happen right now :-)